src/shared/lib/GenericAssayUtils/GenericAssayCommonUtils.ts

```diff
@@ -1 +1,132 @@
+        );
+    });
+
+    const genericAssayMetaGroupByMolecularProfileId: {
+        [molecularProfileId: string]: GenericAssayMeta[];
+    } = {};
+
+    await Promise.all(
+        genericAssayProfiles.map(async genericAssayProfile => {
+            const genericAssayProfileId =
+                genericAssayProfile.molecularProfileId;
+            genericAssayMetaGroupByMolecularProfileId[
+                genericAssayProfileId
+            ] = await fetchGenericAssayMetaByProfileIds([
+                genericAssayProfileId,
+            ]);
+        })
+    );
+    return genericAssayMetaGroupByMolecularProfileId;
+}
+
+export async function fetchGenericAssayMetaByProfileIds(
+    genericAssayProfileIds: string[]
+) {
+    let genericAssayMeta: GenericAssayMeta[] = [];
+    if (genericAssayProfileIds.length > 0) {
+        genericAssayMeta = await client.fetchGenericAssayMetaDataUsingPOST({
+            genericAssayMetaFilter: {
+                molecularProfileIds: genericAssayProfileIds,
+                // the Swagger-generated type expected by the client method below
+                // incorrectly requires both molecularProfileIds and genericAssayStableIds;
+                // use 'as' to tell TypeScript that this object really does fit.
+            } as GenericAssayMetaFilter,
+        });
+    }
+    return genericAssayMeta;
+}
+
+export async function fetchGenericAssayMetaByEntityIds(entityIds: string[]) {
+    let genericAssayMeta: GenericAssayMeta[] = [];
+    if (entityIds.length > 0) {
+        genericAssayMeta = await client.fetchGenericAssayMetaDataUsingPOST({
+            genericAssayMetaFilter: {
+                genericAssayStableIds: entityIds,
+                // the Swagger-generated type expected by the client method below
+                // incorrectly requires both molecularProfileIds and genericAssayStableIds;
+                // use 'as' to tell TypeScript that this object really does fit.
+            } as GenericAssayMetaFilter,
+        });
+    }
+    return genericAssayMeta;
+}
+
+export async function fetchGenericAssayData(
+    entityIdsByProfile: { [molecularProfileId: string]: string[] },
+    sampleFilterByProfile: { [molecularProfileId: string]: IDataQueryFilter }
+) {
+    const params = Object.keys(entityIdsByProfile).map(profileId => ({
+        molecularProfileId: profileId,
+        genericAssayDataFilter: {
+            genericAssayStableIds: entityIdsByProfile[profileId],
+            ...sampleFilterByProfile[profileId],
+            // the Swagger-generated type expected by the client method below
+            // incorrectly requires both samples and a sample list;
+            // use 'as' to tell TypeScript that this object really does fit.
+        } as GenericAssayDataFilter,
+    }));
+    const dataPromises = params.map(param =>
+        client.fetchGenericAssayDataInMolecularProfileUsingPOST(param)
+    );
+    const results: GenericAssayData[][] = await Promise.all(dataPromises);
```

Comment:
if you do this right, TS should be able to infer the return type based on the fact that its an array of whatever is returned by client.fetchGenericAssayDataInMolecularProfileUsingPOST  (Promise<whatever>)

Reply:
@alisman I think the typing for Promise.all is so complicated that the type inference might not work perfectly. might be wrong though.

Reply:
Updated, TS is able to infer the type from the promised result.